assets/js/base/hooks/use-local-storage-state.ts

```diff
@@ -0,0 +1,21 @@
+/**
+ * External dependencies
+ */
+import { useEffect, useState } from '@wordpress/element';
+
+export const useLocalStorageState = < T >(
+	key: string,
+	initialValue: T
+): [ T, ( arg0: T ) => void ] => {
+	const [ state, setState ] = useState< T >( () => {
+		const valueInLocalStorage = window.localStorage.getItem( key );
+		return valueInLocalStorage
+			? JSON.parse( valueInLocalStorage )
+			: initialValue;
+	} );
+	useEffect( () => {
+		window.localStorage.setItem( key, JSON.stringify( state ) );
+	}, [ key, state ] );
+
+	return [ state, setState ];
+};
```

Comment:
Should we add some try catch statements here to be safe with the JSON methods and the local storage operations?

Reply:
Good point, done in 7d810f0c49bdc7d4850dd9307e29bff572b44691.